Initial

aljones15 opened the pull request:
This is the paypal plugin for bedrock-payment.

This might require minor refactoring once Longley finishes the bedrock-payment rewrite.

Addresses: https://github.com/digitalbazaar/bedrock-payment-paypal/issues/1

aljones15 committed:
Add initial files.

Add logger.

Add peerDependencies.

aljones15 committed:
Update peerDeps & use bedrock-payment module.

aljones15 committed:
Update author.

aljones15 committed:
Add license.

aljones15 committed:
Add README.

aljones15 committed:
Use bedrock config for settings.

aljones15 committed:
Further comment on functions.

aljones15 committed:
Add a note on where to find the Payment class.

aljones15 committed:
getOrder can return null.

aljones15 committed:
Move paypal error messages into getConfig.

aljones15 committed:
Add application context to order.

aljones15 committed:
Ensure amounts sent to PayPal are 2 decimal precision.

aljones15 committed:
Validate currencies are supported by PayPal and have the right decimal amount.

aljones15 committed:
Currencies are sets.

aljones15 committed:
Throw if config not found.

aljones15 committed:
Return service name with credentials.

aljones15 committed:
Check for response status.

aljones15 committed:
Await save.

aljones15 committed:
Start abstracting away order.

aljones15 committed:
Do not export unused methods.

aljones15 committed:
Remove unused methods & update plugin api method names.

aljones15 committed:
Update CHANGELOG & add copyrights.

aljones15 committed:
Move axios and bignumber.js to dependencies.

aljones15 committed:
Remove scripts from package.json.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+const bedrock = require('bedrock');
+const BigNumber = require('bignumber.js');
+
+const logger = require('./logger.js');
```
no .js on require statements, although it is there for import statement in es6 modules.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/ded7a74d86b8fe8c87f348c3308380ddbd83b40b

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+const getGatewayCredentials = () => {
+  const {clientId} = getConfig();
+  if(!clientId) {
+    throw new BedrockError(
```
looks to me that `getConfig` is going to throw on L28 if this happens, so this test/error is not required.

aljones15 replied:
Good catch and you are correct. Details here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/ded7a74d86b8fe8c87f348c3308380ddbd83b40b

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+
+  if(!api) {
+    throw new BedrockError(
+      'Missing PayPal api', Errors.Data);
```
```suggestion
      'Missing PayPal API.', Errors.Data);
```

Please add punctuation to other errors.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+    }
+  };
+  const body = 'grant_type=client_credentials';
+  const {data} = await axios.post(authUrl, body, options);
```
I think we would want to catch and rethrow a potential error here, I think it's likely that the extremely verbose axios error likely contains all these secrets and we would want to ensure that info would never be surfaced to a user.

aljones15 replied:
Good idea addressed here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/808ee75f6cc703d5fd13c113bb97965eb80bf99d

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+  } catch(e) {
+    logger.error('getOrder not found', {error: e});
+    if(e.response.status === 404) {
+      return null;
```
Why not throw a NotFoundError.

aljones15 replied:
We need to mark the Payment as Failed and I had issues doing a database save inside of a `try` `catch` loop I'm thinking about how to handle error propagation better.

aljones15 replied:
This is addressed here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/808ee75f6cc703d5fd13c113bb97965eb80bf99d

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+    return data;
+  } catch(e) {
+    logger.error('getOrder not found', {error: e});
+    if(e.response.status === 404) {
```
one of the reasons I took to using apisauce module is because of axios error wonkiness: https://github.com/axios/axios#handling-errors

Unfortunately, you can't count on `e.response` existing and therefore you can hit an error attempting to access `status` of `undefined`.

https://github.com/infinitered/apisauce

aljones15 replied:
I looked into using API Sauce. It looks great, is backwards compatible with axios, etc. however I could not easily find a way of doing old school auth with it. So I'm going to skip on this one unless time allows.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+ * @returns {Promise<object>} The order from the payment gateway.
+ */
+const getOrderFromPayment = async ({payment}) => {
+  const order = await getOrder({id: payment.serviceId});
```
IMO, `getOrder` should throw on notfound.

aljones15 replied:
It does now: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/808ee75f6cc703d5fd13c113bb97965eb80bf99d

I update the Payment status inside of a try catch.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+ *
+ * @returns {Promise<object>} The data returned with an `access_token`.
+ */
+const getAuthToken = async ({clientId, secret, api}) => {
```
Looks like many calls to other APIs lead to a new request for a token.  Do these tokens have a TTL or something, are there any opportunities for cacheing the token instead of adding this round-trip to every request?

aljones15 replied:
Good point. I actually think caching the token here would be better.

aljones15 replied:
ok good idea. Implementation could maybe use some work. Let me know: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/9f162230b10c3c0a9d15a79784ef35c417eea2a4

gannan08 replied:
@mattcollier Did the https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/9f162230b10c3c0a9d15a79784ef35c417eea2a4 commit address your concerns?

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+};
+
+/**
+ * Delete order only works on orders where no money has been exchanged.
```
If the status is not one of the required types do they return something other than a 200 response?  Otherwise, should data be inspected here to determine success/failure?  We probably don't want clients/users interacting with raw paypal responses.

aljones15 replied:
Not sure I'm completely following this. You are asking what status code is returned if the PayPal order status is not correct? Also I believe this will throw if it's not successful.

mattcollier replied:
When you say "Delete order only works", what does that mean?  Does that mean that the paypal APIs are expected to produce a non 200 status code if those conditions are not met?  Are we handling those cases properly?

aljones15 replied:
you are correct. PayPal's API will throw if the order status is not `CREATED` or `AUTHORIZED`.
This is only called by processGateWayPayment if the PayPal order's status is `CREATED`. 
It could throw if somehow the status or the order were to change before the delete action happens. In that case deleteOrder will throw then processGateWayPayment will throw, but the payment can be resumed later. If on resume the PayPal order is now set to COMPLETED the order will process as normally, other wise it will be marked as FAILED. It's not possible to change the PayPal order status after it leaves the CREATED or AUTHORIZED statuses.

CREATED or AUTHORIZED PayPal statuses (these are different than our PaymentStauts enum) mean no money has exchanged hands yet. No credit card has been charged, no PayPal credit used, etc.

gannan08 replied:
Forgive me if it feels like I am repeating things already covered. I am still unclear.

> "Delete order only works on orders where no money has been exchanged."

1. Who controls whether or not "delete order" works? Paypal's code or our code?
2. It doesn't make sense when you say PayPal's API throws an error since it's an HTTP Request. What HTTP Error Codes are thrown when the conditions are not met?
3. We have a single catch-all for any type of error and then we throw. Should all errors be treated the same?

aljones15 replied:
@gannan08 

1. papypal controls which delete operations fail:

> Cancels an order, by ID, and deletes the order. To call this method, the order status must be CREATED or APPROVED.

2. I would consider HTTP status codes from an API endpoint that are above the 400 level to be an error.

https://developer.paypal.com/docs/api/orders/v1/#orders-cancel-response
> A successful request returns the HTTP 204 No Content status code with no JSON response body. If the order is already paid, the order cannot be canceled and the request returns the HTTP 422 Unprocessable Entity status code with the message, This order is in progress.

3. We have to treat axios errors differently because they contain the Authorization header with an oath token that lasts an hour. Hence it would be possible to use our paypal oauth token for an hour if one were to get it from the logs. So not so sure on a one size fits all solution for errors.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+  // at this stage so let's just cancel it.
+  if(order.status === 'CREATED') {
+    await deleteOrder({order});
+    return {};
```
Seems like this is some unexpected state and an error should be thrown after doing the cleanup.

aljones15 replied:
this propagates up to the HTTP Handler which then throws there. I'm not sure what I think about this design either.

aljones15 replied:
Addressed here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/afa1265a02cffe57b8c23a27a2045e62c4b1f466

I added `void: true` to details and then in bedrock-payment-http if an error has `void:true` in details we void the payment there.

aljones15 replied:
Just so you know this was further revised. I pass payment into verifyOrder now so I can set the payment.status inside of the plugin and not in bedrock-payment-http.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+  const verifiedPurchase = verifyPurchase({paypalPurchases});
+  // Guards against the user re-using a previous order from PayPal.
+  const existingPayments = await paymentService.db.findAll(
+    {query: {service: 'paypal', serviceId: order.id}});
```
`serviceId` seems like a strange name in the data model for an `order.id`

aljones15 replied:
in this case `order` refers to a PayPal order hence why it is being assigned to serviceId. I guess this is a case of PayPal's naming conventions conflicting with our naming conventions. For stripe serviceId would be the charge made for instance. We happened to have agreed on the terms Payment and Order for our payments system and it turns out PayPal uses similar terms. orders inside of this module only refer to PayPal orders.

mattcollier commented on index.js:
```diff
@@ -0,0 +1,414 @@
+};
+
+module.exports = {
+  type: 'paymentsPayPalPlugin',
```
maybe type = 'paymentsPlugin' and id = 'paypal'.  Mainly I think these type of plugins should all have the same type and some other way of differentiating the name.  I don't know what pattern is being used here, but this is what we did elsewhere: https://github.com/digitalbazaar/bedrock-ledger-storage-mongodb/blob/master/lib/index.js#L49

aljones15 replied:
Good idea so I did that here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/ded7a74d86b8fe8c87f348c3308380ddbd83b40b

mattcollier reviewed (changes requested):
I'm concerned about axios errors being thrown throughout here that may contain sensitive authentication information or what have you if those error were to bubble up through the stack into a public facing HTTP API.  

Was there discussion about using the official paypal SDK? https://github.com/paypal/PayPal-node-SDK

aljones15 committed:
Capitalize API.

Co-Authored-By: mattcollier <902041+mattcollier@users.noreply.github.com>

aljones15 committed:
Set plugin id and type to paymentPlugin.

aljones15 committed:
Do not reveal PayPal Auth info if the authorization request errors.

aljones15 committed:
Throw an Error if order is CREATED.

aljones15 committed:
Change payment status inside of the Payment Plugin.

aljones15 committed:
Use cached auth data for 45 minutes.

aljones15 committed:
Use camelCase for paypalClientId.

aljones15 committed:
Use node-cache for PayPal data.

aljones15 committed:
Expire cache 5 seconds befor PayPal auth token TTL.

aljones15 committed:
Format Axios errors to prevent leaking auth tokens to the logger.

aljones15 commented:
Refactored and this is what the message looks like now:

```
 "error": {
    "name": "NotAllowedError",
    "message": "PayPal Authentication failed.",
    "details": {
      "httpStatusCode": 401,
      "name": "Error",
      "message": "Request failed with status code 401"
    },
    "cause": null
  }
```

aljones15 committed:
Remove cause from formatAxiosError.

mattcollier commented:
@aljones15 I see your latest iteration regarding error messages.  I think I still prefer the solution that I suggested earlier: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2#commitcomment-35873470

Here's where I'm coming from.  I thing surfacing HTTP errors from these plugins to the payments API is a fail.  The main function of these plugin APIs is to abstract away the HTTP interface from every payment processor.  The aim being that from the perspective of the Payments API that calls the plugin, calling the Stripe plugin and the Paypal plugin should have exactly the same API surface which would include the errors being returned.  Therefore, this API needs to wrap all the axios errors in some consistent way, not allowing Axios specific errors/error types to bubble up the stack.

So, throwing consistent BedrockErrors and specifying the summary of the axios error as a `cause` as I suggested earlier provides for this.

Another way to look at it would be to imagine that the future Stripe plugin might utilize an entirely different http client API.

What is really being communicated from these plugins to the payments API in the case of failure?  The Payments API wants to know if the request failed due to Paypal saying it failed. (NotFoundError, NotAllowedError) in which case, resending the same request again would not produce success.  Or, was it a temporary network failure of some kind, (Timeout, Host could not be reached), where retrying the request again might produce success.

I feel like the error handling provided by `apisauce` goes a long way towards addressing some of these concerns: https://github.com/infinitered/apisauce/blob/master/lib/apisauce.ts#L79-L124

aljones15 committed:
Use new data model.

aljones15 committed:
formatAxiosError just formats an axios error.

aljones15 committed:
Add a period to the end of all BedrockErrors.

aljones15 committed:
Make all Errors Human Readable.

aljones15 committed:
Add errorTypes based off of PayPal status codes.

aljones15 committed:
Use formatAxios Error.

gannan08 reviewed (changes requested):
Reading this code could be quite difficult for developers. My first read made the assumption that these cases were unfinished and you were coming back to fill them in later, I was wrong. My second read made the assumption that if the case didn't have anything then it would fall to default, that's wrong too. What you're actually doing here is using the fall-through aspect of switch cases to return errors. You're also not using `break`, though not required, it is a statement one would expect with switches like peanut butter and jelly.

My read is that you are grouping errors into a set and IF an error status is an element of that set THEN return an error associated with that set. That could be achieved with the following code:

```js
const notAllowedErrors = [401, 403];

if(notAllowedErrors.includes(status)) {
  return {details, errorType: Errors.NotAllowed};
}

```


I believe the if/else pattern describes the intention of the code and provides better readability here. Just because one could code a solution one way and it is syntactically correct does not mean one should as there are other factors that go into play for writing code.

aljones15 committed:
Use sets for the status code errors.

mattcollier reviewed (commented):
Nice!

aljones15 committed:
Update getOrderPayment arguements.

aljones15 committed:
Ensure currencies are compared.

aljones15 committed:
Add bedrock config.

aljones15 committed:
Start tests.

aljones15 committed:
Add update tests.

aljones15 committed:
Add periods to the end of test names.

aljones15 committed:
Add copyrights to test files.

  - Remove license from package.json.
  - Add private: true to package.json.
  - Remove should from tests.
  - Add eslint json file.

aljones15 committed:
encodeURIComponent on order.id.

  - Add a test cards file.
  - Add a processor method for tests.
  - Add more process tests.

aljones15 committed:
use puppeteer to fill in payment information.

aljones15 committed:
Add processing tests.

aljones15 committed:
Add explicit timeouts to all tests.

aljones15 committed:
Make config errors more descriptive.

aljones15 committed:
Move capture payment into helpers.

aljones15 committed:
Use nock for all processing tests.

  - Add a mock-paypal file to create mock paypal responses.
  - Add nock to package.json

aljones15 committed:
Assert on errors in Process.

aljones15 committed:
Change first tests over to nock.

aljones15 committed:
Create integration test dir.

aljones15 committed:
Add integration tests.

aljones15 committed:
Add generic stubs for paypal api calls.

aljones15 committed:
Use nock for all tests.

aljones15 committed:
Enable running integration tests.

aljones15 committed:
Add copyrights to test files.

aljones15 committed:
Move nock to dependencies.

aljones15 committed:
Update test project name in package.json.

Co-Authored-By: mattcollier <902041+mattcollier@users.noreply.github.com>

aljones15 committed:
Remove license & add white space below copyright notice.

aljones15 committed:
Make config errors more general.

davidlehn commented on index.js:
```diff
@@ -0,0 +1,555 @@
+const getOptions = async ({clientId, secret, api} = getConfig()) => {
+  const {access_token, token_type} = await getAuthToken(
+    {clientId, secret, api});
+  const Authorization = `${token_type} ${access_token}`;
```
Could inline this in the headers object.

aljones15 replied:
good suggestion done here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/90f3edaba533e086ea7c70567f8e0f0c0e4c65ca

Thanks.

davidlehn commented on package.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "bedrock-payment-paypal",
+  "version": "0.0.1",
+  "description": "This Plugin provides PayPal functionality to bedrock-payment.",
```
Add license field.

aljones15 replied:
Are you sure about this? I thought if there is a License.md in the repo we don't need a license in the package.json? Also what is the license type for the field? Usually use BSD-3-Clause, but this uses the bedrock license.

davidlehn replied:
npm warns about no license on install.  And there are tools to recusively report all licenses used in a project via that field.  We've been using what they suggest, which looks absurd and renders poorly on their own site:
https://docs.npmjs.com/files/package.json#license
`"license" : "SEE LICENSE IN LICENSE.md"`

aljones15 replied:
done here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/f8d86f95a4c7aeb5508965a7b184d703b8bcf9a9

davidlehn commented on package.json:
```diff
@@ -0,0 +1,44 @@
+    "eslint-plugin-jsdoc": "^17.1.0"
+  },
+  "peerDependencies": {
+    "bedrock": "^2.0.0",
```
```suggestion
    "bedrock": "2.x - 3.x",
```

davidlehn commented on test/package.json:
```diff
@@ -0,0 +1,32 @@
+  "scripts": {
+    "test": "node --preserve-symlinks test test"
+  },
+  "keywords": [
```
Can simplify and remove the fields that are not needed for tests to run.  The ones in the top level package.json are the ones that matter.

aljones15 replied:
While I will remember this in the future... fondly... do I really need to eliminate more identifying information from a package.json?

davidlehn replied:
Don't need to, just suggesting how to simplify away unused fields.

aljones15 replied:
Going to skip on this one then. I did refactor the errors: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/25bcd406c6c49b9ee691a4f748e1367c851ef25e

davidlehn commented on README.md:
```diff
@@ -0,0 +1,67 @@
+
+👤 **Digital Bazaar, Inc.**
+
+* Github: [@digitalbazaar](https://github.com/digitalbazaar)
```
```suggestion
* GitHub: [@digitalbazaar](https://github.com/digitalbazaar)
```

aljones15 committed:
Update bedrock version in package.json.

Co-Authored-By: David I. Lehn <dil@lehn.org>

aljones15 committed:
Add github link to README.

Co-Authored-By: David I. Lehn <dil@lehn.org>

aljones15 committed:
Inline auth header in headers.

aljones15 committed:
Add license to bedrock-payment-paypal.

aljones15 committed:
Use details instead of strings.

aljones15 committed:
Add paypal id to currency and amount errors.

davidlehn commented on README.md:
```diff
@@ -0,0 +1,67 @@
+Give a ⭐️ if this project helped you!
+
+***
+_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_
```
Some pushback on using this tool:
- This README is different than every other project we have.
- Why is there HTML output from a 'md' generator?
- The version badge won't auto-update.  We should avoid using things like that.
- I guess the emoji use is cute?  Seems like a bit much.

Perhaps we could just use the similar but simpler https://github.com/RichardLitt/standard-readme style like in https://github.com/digitalbazaar/crypto-ld?

aljones15 replied:
@davidlehn that guys a yeoman template for standard read me generation: https://github.com/RichardLitt/generator-standard-readme

aljones15 replied:
@davidlehn  I removed the html parts and I installed the standard read me generator. I'm not sure if I will get time to refactor this README, but I am a yeoman template away from it in the future. Removed the HTML here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/3da073cedb3ac354b32ddc7343c6996e66046149

davidlehn replied:
I think copy and and modify with any editor can get close enough.  No reason to involve generators or other tools for a readme.

aljones15 replied:
@davidlehn used the generator and added a standard README: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/79df045d67ce18b90272d328b40c102113ca9b57

aljones15 committed:
Remove HTML from README.

aljones15 committed:
Add a standard README to the project.

gannan08 commented on config.js:
```diff
@@ -0,0 +1,16 @@
+const c = util.config.main;
+const cc = c.computer();
+
+cc('paypal', {
```
Why are we not naming this configuration `bedrock-payment-paypal`? The convention for config namespace names have usually been their package name. Why are we using a computed config here? I'd assume we use computed configs when these values are dynamic and may change. Please correct me if I'm wrong @davidlehn.

davidlehn replied:
Yes, we usually use package name.  Though sometimes omit the 'bedrock-' prefix.

Doesn't look like computed config is needed here since it's just initializing fields.

aljones15 replied:
Yeah we went with `payment-http` elsewhere. This is now `bedrock-payment-paypal` https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

anyways will try to use the package name in the future.

davidlehn replied:
Could also remove the computed config part.  But no big deal either way.

aljones15 replied:
I would prefer to stick with just using computed for all settings. I seem to have issues figuring out when to use computed and when to not so I'm just going to use `cc` whenever I can.

gannan08 commented on index.js:
```diff
@@ -0,0 +1,563 @@
+    createGatewayPayment,
+    processGatewayPayment
+  },
+  test: {
```
Don't think we should be combining code for testing infrastructure in the main module. We have avoided this in many of our libraries and we should continue.

aljones15 replied:
unfortunately most of the methods this depends on are private to this module so we have 2 options:

1. expose private methods in the plugin not used by anything else except one test method.
2. rewrite all the dependent private methods in a test utils file (copy and paste perhaps).

gannan08 commented:
In this module it seems that we have the concept of an order. Is it true that we are referring to both a bedrock order or a paypal order? If that is the case then we have unfortunately overloaded the term and it makes the conceptual understanding of this library a little more difficult to grasp. Can we be intentional with naming here: call things either a "Bedrock Order/bedrockOrder" or "PayPal Order/paypalOrder". This may or may not be a problem for others, but it did trip me up multiple times.

davidlehn commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+.cache
+.project
+.settings
+*.log
```
Sort at top.

aljones15 replied:
Sorted here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on CHANGELOG.md:
```diff
@@ -0,0 +1,10 @@
+## 0.0.1 -
+
+### Added
+  - Initial index.js.
```
Remove indents.
Do these lines cover the whole commit or would general shorter comment work better here?

aljones15 replied:
Removed: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on README.md:
```diff
@@ -0,0 +1,95 @@
+
+## License
+
+Bedrock License © 2019 digitalbazaar
```
```suggestion
Bedrock License © 2019 Digital Bazaar, Inc.
```

davidlehn commented on index.js:
```diff
@@ -0,0 +1,563 @@
+ * and match the expectedAmount.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.payment - A Bedrock payment.
```
options don't match code.

aljones15 replied:
Updated here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on index.js:
```diff
@@ -0,0 +1,563 @@
+  const updatedOrder = await updateOrder({order, patch});
+  // make sure the swap occurred.
+  compareAmount({order: updatedOrder, expectedAmount: amount});
+  const payment = Object.assign(pendingPayment, updatedPayment);
```
Better to not mutate `pendingPayment`.  Just create a new object.
```suggestion
  const payment = Object.assign({}, pendingPayment, updatedPayment);
```

aljones15 replied:
good idea and done. I did not commit your suggestion because I wanted to test it locally first before committing it to the repo. Added it here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on index.js:
```diff
@@ -0,0 +1,563 @@
+ * a valid PayPal Order.
+ *
+ * @param {object} options - Options to use.
+ * @param {Array<object>} options.clientPurchases - Client purchases.
```
Not used?

aljones15 replied:
yeah was an option from a previous revision removing it. https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on index.js:
```diff
@@ -0,0 +1,563 @@
+    throw new BedrockError('Missing PayPal purchase.', Errors.Data);
+  }
+  // throw an error if the client altered their purchase amount
+  // after the transaction succeeded.
```
Not sure what these comments mean.

aljones15 replied:
they were comments for a line which was removed. my apologies removing it. https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on index.js:
```diff
@@ -0,0 +1,563 @@
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - A PayPal order object.
+ * @param {object} options.card - The payee's card.
```
Not used.  Should doc body?

aljones15 replied:
yeah changed it to doc body here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/42a81daf80556f5ef82c3744e34cb7533e466d79

davidlehn commented on test/integration/020-create.js:
```diff
@@ -0,0 +1,77 @@
+
+const {BedrockError} = util;
+const minute = 60000;
+const twoMinutes = minute * 2;
```
Maybe overkill for these testing cases, but the `ms` package makes this easier like `.timeout(ms('2m'))`.

aljones15 replied:
a bit of an overkill for test code.

aljones15 committed:
Use the full Digital Bazaar name in copyright.

Co-Authored-By: David I. Lehn <dil@lehn.org>

aljones15 committed:
Use bedrock-payment-paypal for configuration key.

  - Sort .gitignore.
  - Remove indents from CHANGELOG.
  - Update various doc strings.
